js/imageeditor/shared/layers/Layers.svelte

```diff
@@ -0,0 +1,193 @@
+<script lang="ts">
+	import { createEventDispatcher, getContext, onMount, tick } from "svelte";
+	import { DropdownArrow } from "@gradio/icons";
+	import { click_outside } from "../utils/events";
+	import { layer_manager, type LayerScene } from "./utils";
+	import { EDITOR_KEY, type EditorContext } from "../ImageEditor.svelte";
+	import type { FileData } from "@gradio/client";
+
+	let show_layers = false;
+
+	export let layer_files: (FileData | null)[] | null = [];
+
+	const dispatch = createEventDispatcher<{
+		new_layer: void;
+	}>();
+
+	const { pixi, current_layer, dimensions } =
+		getContext<EditorContext>(EDITOR_KEY);
+
+	const LayerManager = layer_manager();
+	let layers: LayerScene[] = [];
+
+	function once(fn: () => void): () => void {
+		let called = false;
+		return () => {
+			if (called) return;
+			called = true;
+			fn();
+		};
+	}
+	const once_layer = once(new_layer);
+
+	$: $pixi && once_layer();
+
+	async function new_layer(): Promise<void> {
+		if (!$pixi) return;
+
+		await tick();
+
+		setTimeout(() => {
+			const [active_layer, all_layers] = LayerManager.add_layer(
+				$pixi.layer_container,
+				$pixi.renderer,
+				...$dimensions
+			);
+
+			$current_layer = active_layer;
+			layers = all_layers;
+		}, 20);
+	}
+
+	$: render_layer_files(layer_files);
+
+	function is_not_null<T>(x: T | null): x is T {
+		return x !== null;
+	}
+
+	async function render_layer_files(
+		_layer_files: typeof layer_files
+	): Promise<void> {
+		await tick();
+		if (!_layer_files || _layer_files.length == 0) return;
+		if (!$pixi) return;
+
+		const fetch_promises = await Promise.all(
+			_layer_files.map((f) => {
+				if (!f || !f.url) return null;
+
+				return fetch(f.url);
+			})
+		);
+
+		const blobs = await Promise.all(
+			fetch_promises.map((p) => {
+				if (!p) return null;
+				return p.blob();
+			})
+		);
+
+		LayerManager.reset();
+
+		let last_layer: [LayerScene, LayerScene[]];
+		for (const blob of blobs.filter(is_not_null)) {
+			last_layer = await LayerManager.add_layer_from_blob(
+				$pixi.layer_container,
+				$pixi.renderer,
+				blob
+			);
+		}
+
+		$current_layer = last_layer[0];
+		layers = last_layer[1];
+	}
+
+	onMount(async () => {
+		await tick();
+		if (!$pixi) return;
+		function reset(): void {
+			console.log("resetting");
+			LayerManager.reset();
+			$pixi?.resize(...$dimensions);
+
+			new_layer();
+		}
+		$pixi = { ...$pixi!, reset, get_layers: LayerManager.get_layers };
+	});
```

Comment:
I think this is the problem. We do this in a few places and it suck's because we can't guarantee the order that the various reset functions are called. 

Instead we should dispatch events from the core component when we want to reset something and manually call reset at a higher level. Alternatively we can track functions more granularly.